src/cls/rgw/cls_rgw_types.h

```diff
@@ -352,83 +352,7 @@ struct rgw_bucket_entry_ver {
 };
 WRITE_CLASS_ENCODER(rgw_bucket_entry_ver)
 
-
-struct cls_rgw_obj_key {
-  std::string name;
-  std::string instance;
-
-  cls_rgw_obj_key() {}
-  cls_rgw_obj_key(const std::string &_name) : name(_name) {}
-  cls_rgw_obj_key(const std::string& n, const std::string& i) : name(n), instance(i) {}
-
-  std::string to_string() const {
-    return fmt::format("{}({})", name, instance);
-  }
-
-  bool empty() const {
-    return name.empty();
-  }
-
-  void set(const std::string& _name) {
-    name = _name;
-    instance.clear();
-  }
-
-  bool operator==(const cls_rgw_obj_key& k) const {
-    return (name.compare(k.name) == 0) &&
-           (instance.compare(k.instance) == 0);
-  }
-
-  bool operator!=(const cls_rgw_obj_key& k) const {
-    return (name.compare(k.name) != 0) ||
-           (instance.compare(k.instance) != 0);
-  }
-
-  bool operator<(const cls_rgw_obj_key& k) const {
-    int r = name.compare(k.name);
-    if (r == 0) {
-      r = instance.compare(k.instance);
-    }
-    return (r < 0);
-  }
-
-  bool operator<=(const cls_rgw_obj_key& k) const {
-    return !(k < *this);
-  }
-
-  void encode(ceph::buffer::list &bl) const {
-    ENCODE_START(1, 1, bl);
-    encode(name, bl);
-    encode(instance, bl);
-    ENCODE_FINISH(bl);
-  }
-  void decode(ceph::buffer::list::const_iterator &bl) {
-    DECODE_START(1, bl);
-    decode(name, bl);
-    decode(instance, bl);
-    DECODE_FINISH(bl);
-  }
-  void dump(ceph::Formatter *f) const {
-    f->dump_string("name", name);
-    f->dump_string("instance", instance);
-  }
-  void decode_json(JSONObj *obj);
-  static void generate_test_instances(std::list<cls_rgw_obj_key*>& ls) {
-    ls.push_back(new cls_rgw_obj_key);
-    ls.push_back(new cls_rgw_obj_key);
-    ls.back()->name = "name";
-    ls.back()->instance = "instance";
-  }
-
-  size_t estimate_encoded_size() const {
-    constexpr size_t start_overhead = sizeof(__u8) + sizeof(__u8) + sizeof(ceph_le32); // version and length prefix
-    constexpr size_t string_overhead = sizeof(__u32); // strings are encoded with 32-bit length prefix
-    return start_overhead +
-        string_overhead + name.size() +
-        string_overhead + instance.size();
-  }
-};
-WRITE_CLASS_ENCODER(cls_rgw_obj_key)
+typedef rgw_obj_index_key cls_rgw_obj_key;
```

Comment:
can cls_rgw_types.h use the more granular includes instead of pulling in everything from rgw_basic_types.h? are rgw_obj_types.h and rgw_user_types.h enough?

Reply:
My intent was to make the cleanup 100% self-testing--because by including the batched header, we ensure that the whole thing compiles in cls context.

Reply:
Because, yes, otherwise, it violates include-what-you use that I was just griping about :)

Reply:
> My intent was to make the cleanup 100% self-testing--because by including the batched header, we ensure that the whole thing compiles in cls context.

understood. i just don't think we *want* to share all of these basic-type headers between rgw, sal, and cls. the types shared between rgw and cls_rgw are rados-specific, so we don't want them to be visible to sal at all

rgw depends on cls_rgw_client, so this inclusion of rgw_basic_types.h is already a circular dependency. we shouldn't expose more rgw types here unnecessarily, or we risk accidentally adding new dependencies on them in cls_rgw. for the types that do need to be shared, i would rather move them into cls_rgw_types.h and let rgw_sal_rados include them from there

it's fine if this stuff is out-of-scope for the MPU work, i just wanted to raise it for further zipper discussion

Reply:
well, I guess you make a good point about SAL, I haven't thought clearly enough about that.
but the circular dependency between cls_tyes and rgw_basic_types isn't really there, as we've proved by construction.  I think I generally disagree that that should be the relationship--allowing type definitions to be cleanly shared between at least rgw and cls_rgw is what this change does.